File: src/mpmm.cpp
#include <TMB.hpp>

using namespace density;

// modified from glmmTMB
// need to link to these via glmmTMB pkg but can't currently because they do not export C++
template <class Type>
struct per_term_info {
  // Input from R
  int blockCode;     // Code that defines structure
  int blockSize;     // Size of one block
  int blockReps;     // Repeat block number of times
  int blockNumTheta; // Parameter count per block
  matrix<Type> dist;
  // Report output
  matrix<Type> corr;
  vector<Type> sd;
};
// need to link to these via glmmTMB pkg
template <class Type>
struct terms_t : vector<per_term_info<Type> > {
  terms_t(SEXP x){
    (*this).resize(LENGTH(x));
    for(int i=0; i<LENGTH(x); i++){
      SEXP y = VECTOR_ELT(x, i);    // y = x[[i]]
      int blockCode = (int) REAL(getListElement(y, "blockCode", &isNumericScalar))[0];
      int blockSize = (int) REAL(getListElement(y, "blockSize", &isNumericScalar))[0];
      int blockReps = (int) REAL(getListElement(y, "blockReps", &isNumericScalar))[0];
      int blockNumTheta = (int) REAL(getListElement(y, "blockNumTheta", &isNumericScalar))[0];
      (*this)(i).blockCode = blockCode;
      (*this)(i).blockSize = blockSize;
      (*this)(i).blockReps = blockReps;
      (*this)(i).blockNumTheta = blockNumTheta;
    }
  }
};

enum valid_covStruct {
  diag_covstruct = 0,
  us_covstruct   = 1
};
// would be better to have AR(1) corrStruct here to deal with serial dependence in animal tracks
// need to link to these via glmmTMB pkg but can't currently because they do not export C++
template <class Type>
Type termwise_nll(array<Type> &U, vector<Type> theta, per_term_info<Type>& term) {
  Type ans = 0;
  if (term.blockCode == diag_covstruct){
    // case: diag_covstruct
    vector<Type> sd = exp(theta);
    for(int i = 0; i < term.blockReps; i++){
      ans -= dnorm(vector<Type>(U.col(i)), Type(0), sd, true).sum();
    }
    term.sd = sd;
  }
  else if (term.blockCode == us_covstruct){
    // case: us_covstruct
    int n = term.blockSize;
    vector<Type> logsd = theta.head(n);
    vector<Type> corr_transf = theta.tail(theta.size() - n);
    vector<Type> sd = exp(logsd);
    density::UNSTRUCTURED_CORR_t<Type> nldens(corr_transf);
    density::VECSCALE_t<density::UNSTRUCTURED_CORR_t<Type> > scnldens = density::VECSCALE(nldens, sd);
    for(int i = 0; i < term.blockReps; i++){
      ans += scnldens(U.col(i));
    }
    term.corr = nldens.cov();
    term.sd = sd;
  }
  else error("covStruct not implemented!");
  return ans;
}
// need to link to these via glmmTMB pkg but can't currently because they do not export C++
template <class Type>
Type allterms_nll(vector<Type> &u, vector<Type> theta,
                  vector<per_term_info<Type> >& terms) {
  Type ans = 0;
  int upointer = 0;
  int tpointer = 0;
  int nr, np = 0, offset;
  for(int i=0; i < terms.size(); i++){
    nr = terms(i).blockSize * terms(i).blockReps;
    // Note: 'blockNumTheta=0' ==> Same parameters as previous term.
    bool emptyTheta = ( terms(i).blockNumTheta == 0 );
    offset = ( emptyTheta ? -np : 0 );
    np     = ( emptyTheta ?  np : terms(i).blockNumTheta );
    vector<int> dim(2);
    dim << terms(i).blockSize, terms(i).blockReps;
    array<Type> useg( &u(upointer), dim);
    vector<Type> tseg = theta.segment(tpointer + offset, np);
    ans += termwise_nll(useg, tseg, terms(i));
    upointer += nr;
    tpointer += terms(i).blockNumTheta;
  }
  return ans;
}

// mpmm model code
template<class Type>
Type objective_function<Type>::operator() ()
{

  DATA_MATRIX(X);                   // fixed effects design matrix
  DATA_SPARSE_MATRIX(Z);            // random effects design sparse matrix
  DATA_MATRIX(xy);                  // locations
  DATA_FACTOR(idx);                 // cumsum of number of locations for each animal
//  DATA_FACTOR(didx);                // cumsum of number of dive observations for each animal
  DATA_VECTOR(d);                   // dive summary statistic
  DATA_VECTOR(dt);                  // time interval b/w dive observations
  DATA_IVECTOR(obs);                // indicator for missing data

  // Define covariance structure for the conditional model
  DATA_STRUCT(terms, terms_t);
  DATA_INTEGER(A);                // number of animals

  PARAMETER_VECTOR(lg);		          // move autocorrelation parameter (gamma on link scale)
  PARAMETER_VECTOR(ll);             // dive autocorrelation parameter
  PARAMETER_VECTOR(beta);           // fixed regression coefficients (link scale)
  PARAMETER_VECTOR(b);              // random intercept & slope terms
  PARAMETER_VECTOR(log_sigma);	    // innovation variance (log scale)
  PARAMETER(log_sigma_g);           // logistic scale parameter of rw on lg (log scale)
  PARAMETER(log_sigma_l);
  PARAMETER_VECTOR(theta);          // covariance parameters


// Backtransform parameters from link scale
vector<Type> gamma = Type(1.0) / (Type(1.0) + exp(-lg));
vector<Type> lambda = Type(1.0) / (Type(1.0) + exp(-ll));
vector<Type> sigma = exp(log_sigma);
Type sigma_g = exp(log_sigma_g);
Type sigma_l = exp(log_sigma_l);

// 2x2 covariance matrix for innovations
matrix<Type> cov(2, 2);
cov(0,0) = sigma(0) * sigma(0);
cov(0,1) = 0.0;
cov(1,0) = 0.0;
cov(1,1) = sigma(1) * sigma(1);

Type jnll = 0.0;
vector<Type> mu(2);

MVNORM_t<Type> nll_dens(cov);
int i,j;

// Random intercept & slope(s)
jnll += allterms_nll(b, theta, terms);

// linear predictor
vector<Type> eta = X * beta + Z * b;

  for(i = 0; i < A; ++i) {
    for(j = idx(i); j < idx(i+1); ++j) {
      jnll -= dnorm(lg(j), eta(j), sigma_g, true);
    }
    for(j = (idx(i)+1); j < idx(i+1); ++j){
      jnll -= dnorm(ll(j), ll(j-1), sigma_l, true);
    }
// CT code    
/*    for(j = (didx(i)+1); j < didx(i+1); ++j) {
//      Type sig_ldt = pow(dt(j) * dt(j) * sigma_l * sigma_l, 0.5);
//      jnll -= dnorm(ll(j), ll(j-1), sig_ldt, true);  // RW on logit(lambda)
    }
*/
    for(j = (idx(i)+2); j < idx(i+1); ++j){
      mu = xy.row(j) - xy.row(j-1) - gamma(j-1) * (xy.row(j-1) - xy.row(j-2));  // first diff RW on locations
      jnll += nll_dens(mu);
    }
    for(j = (idx(i)+1); j < idx(i+1); ++j){
      if(obs(j-1) == 1 && obs(j) == 1) {
        //Type sig_dt = pow(dt(j) * dt(j) * sigma(2) * sigma(2), 0.5);  // CT code
        jnll -= dnorm(d(j), lambda(j-1) * d(j-1), sigma(2), true);
      }
    }

// CT code      
/*    for(j = (didx(i)+1); j < didx(i+1); ++j) {
//      Type sig_dt = pow(dt(j) * dt(j) * sigma(2) * sigma(2), 0.5);
//      jnll -= dnorm(d(j), lambda(j-1) * d(j-1), sigma(2), true);
    }
*/ 
  }

  // Report variance components
  vector<matrix<Type> > corr(terms.size());
  vector<vector<Type> > sd(terms.size());
  for(int i=0; i<terms.size(); i++){
    if(terms(i).blockNumTheta > 0){
      corr(i) = terms(i).corr;
      sd(i) = terms(i).sd;
    }
  }

  REPORT(corr);
  REPORT(sd);
  ADREPORT(sigma_g);
  ADREPORT(sigma_l);
  ADREPORT(sigma);
  ADREPORT(beta);

  return jnll;
}

File: R/mpmm.r
##' Fit a move persistence random walk via TMB to a pre-filtered/regularised animal
##'   track and estimate gamma as a linear function of covariates
##'
##' The input track is given as a dataframe where each row is an
##' observed location and columns
##' \describe{
##' \item{'id'}{individual animal identifier,}
##' \item{'date'}{observation time (POSIXct,GMT),}
##' \item{'lon'}{observed longitude,}
##' \item{'lat'}{observed latitude,}
##' \item{'...'}{named covariates appended to track}
##' }
##'
##' @title Move Persistence Mixed-Effects Model
##' @param formula a right-hand-side regression formula (no response variable)
##' @param data a data frame of observations (see details)
##' @param method method for maximising the log-likelihood ("ML" or "REML")
##' @param optim numerical optimizer to be used (nlminb or optim)
##' @param control a list of control parameters (currently only for nlminb)
##' @param verbose report progress during minimization
##' @return a list with components
##' \item{\code{states}}{a dataframe of estimated states}
##' \item{\code{fitted}}{a dataframe of fitted locations}
##' \item{\code{par}}{model parameter summmary}
##' \item{\code{data}}{input dataframe}
##' \item{\code{tmb}}{the tmb object}
##' \item{\code{opt}}{the object returned by the optimizer}
##' @examples
##' data(ellie.ice)
##' fit <- mpmm(~ ice + (1 | id), data = d.ice)
##' summary(fit)
##'
##' @useDynLib mpmm
##' @importFrom lme4 nobars findbars subbars mkReTrms
##' @importFrom glmmTMB getReStruc splitForm
##' @importFrom Matrix t
##' @importFrom dplyr %>% tbl_df data_frame
##' @importFrom TMB MakeADFun sdreport newtonOption
##' @export
mpmm <- function(
                formula = NA,
                data = NULL,
                method = "ML",
                optim = c("nlminb", "optim"),
                control = NULL,
                verbose = FALSE) {
  
  require(TMB)
  require(glmmTMB)
  require(lme4)
  
  call <- mf <- match.call()
  optim <- match.arg(optim)
  
  compile("src/mpmm.cpp")
  dyn.load(dynlib("src/mpmm"))
  
  # check that the formula is a formula
  is.formula <- function(x)
    tryCatch(
      inherits(x, "formula"),
      error = function(e) {
        FALSE
      }
    )
  if (!is.formula(formula))
    stop("\n'formula' must be specified as ~ x + ...")

  # check that there is no response variable in the formula
  if (attr(terms(formula), "response") != 0)
    stop("\n'formula' can not have a response variable")

  # check that either ML or REML are the specified maximisation method
  if (!method %in% c("ML", "REML"))
    stop("\n'method' argument must be either ML or REML")

  # check that formula has a random component
  if(is.null(findbars(formula)))
    stop("\n formula must include a random component; e.g., ~ (1 | id)")

  # check that covariates do not contain NA's
  covars <- nobars(formula) %>% terms() %>% attr(., "term.labels")
  nas <- is.na(data[, covars]) %>% apply(., 2, sum)
  if (sum(nas) > 0)
    stop(
      paste0(
        "\n NA's detected in the following covariates: ",
        covars[which(nas > 0)],
        "\n consider imputing values"
      )
    )

  # evaluate model frame
  m <- match(c("data", "subset", "na.action"),
             names(mf), 0L)
  mf <- mf[c(1L, m)]
  mf$drop.unused.levels <- TRUE
  mf[[1]] <- as.name("model.frame")
  mf$formula <- formula

  f <- subbars(formula)
  environment(f) <- environment(formula)
  mf$formula <- f
  fr <- eval(mf, envir = environment(formula), enclos = parent.frame())

  # strip random part of formula
  ff <- nobars(formula)
  nobs <- nrow(fr)

  # build fixed effects model matrix X
  # no model matrix if fixed part of formula is empty
  if (identical(ff, ~  0) ||
      identical(ff, ~ -1)) {
    X <- NULL
  } else {
    mf$formula <- ff
    termf <- terms(eval(mf, envir = environment(ff)))
    X <- model.matrix(ff, fr)
    terms <- list(fixed = terms(termf))
  }

  # build random effects sparse matrix Z
  ref <- formula
  if (is.null(findbars(ref))) {
    Z <- matrix(0, nrow = nobs, ncol = 0)
    Z <- as(Z, "dgTMatrix")
    reTrms <- NULL
    ss <- integer(0)
  } else {
    ref <- findbars(ref)
    reTrms <- mkReTrms(ref, fr)
    mf$formula <- ref
    ss <- splitForm(formula)
    ss <- unlist(ss$reTrmClasses)
    Z <- t(reTrms$Zt)
  }

  condList  <- list(X = X, Z = Z, reTrms = reTrms, ss = ss, terms = terms)

  condReStruc <- with(condList, getReStruc(reTrms, ss))

  gnm <- names(condList$reTrms$flist)

  # num observations
  nobs <- nrow(fr)

  # num animals within group
  A <- sapply(reTrms$flist, nlevels)

  # num random effects
  #nre <- sapply(reTrms$cnms, length) %>% sum()

  ## get index of observations per group level - nominally individual animals
  idx <- data$id %>%
    table() %>%
    as.numeric() %>%
    cumsum() %>%
    c(0, .)

  ## create missing observation index for dive stats
  data <- data %>%
    mutate(obs = ifelse(is.na(av.depth), 0, 1)) 
  
  ## calc delta times in hours for continuous-time RW on lambda
  ## define helper function
  dt.fun <- function(x) {
    dt <- difftime(x, lag(x), units = "hours")
    dt <- as.numeric(dt)
    dt[1] <- 0
    dt
  }
    
  dive <- data %>%
    group_by(id) %>%
    select(date, av.depth) %>%
    filter(!is.na(av.depth)) %>%
    mutate(dt = dt.fun(date))
  
  ## get index of irregular dive observations per group level
  ##  needed for the CT version
#  didx <- dive$id %>%
#    table() %>%
#    as.numeric() %>%
#    cumsum() %>%
#    c(0, .)
  
  ## build data for TMB
  data.tmb <- list(
    X     = condList$X,
    Z     = condList$Z,
    xy    = cbind(data$x, data$y),
    idx   = idx,
#    didx  = didx,   # this was needed for the CT version
    d     = ifelse(is.na(data$av.depth), -999, data$av.depth), # set NA's to -999 so TMB is happy
    dt    = dive$dt,
    obs   = data$obs,
    terms = condReStruc,
    A     = A
  )
  
  getVal <- function(obj, component)
    vapply(obj, function(x) x[[component]], numeric(1))

  param <- with(data.tmb,
                     list(
                       lg          = rep(1, nobs),
                       ll          = rep(1, nobs),
                       beta        = rep(1, ncol(X)),
                       b           = rep(1, ncol(Z)),
                       log_sigma   = c(0,0,0),
                       log_sigma_g = 0,
                       log_sigma_l = 0,
                       theta       = rep(0, sum(getVal(condReStruc, "blockNumTheta")))
                     ))
  rnd <- c("lg", "ll", if(ncol(data.tmb$Z) > 0) "b")
  
  forTMB <- list(data = data.tmb,
              param = param,
              rnd = rnd,
              gnm = gnm,
              condList = condList,
              condReStruc = condReStruc,
              allForm = list(formula),
              fr = fr,
              call = call,
              verbose = verbose
              )

  # integrate out the beta's from the likelihood - REML estimation; appends the beta's to the random arg.
  profile <- NULL
  if(method == "REML") profile <- "beta"

  ## TMB - create objective function
  obj <-
    with(
      forTMB,
      MakeADFun(
        data = data.tmb,
        parameters = param,
        random = rnd,
        profile = profile,
        DLL = "mpmm",
      #  hessian = TRUE,
        silent = !verbose
      )
    )

  obj$env$inner.control$trace <- verbose
  obj$env$tracemgc <- verbose

  obj$control <- list(trace = 0,
                      reltol = 1e-12,
                      maxit = 500)
  newtonOption(obj, smartsearch = TRUE)

  ## Minimize objective function
  opt.time <- system.time(opt <- suppressWarnings(switch(
    optim,
    nlminb = nlminb(
      start = obj$par,
      objective = obj$fn,
      gradient = obj$gr,
      control = control
    ),
    optim = do.call("optim", obj)
  )))

  cat(opt.time, "\n")
  
  ## Parameters, states and the fitted values
  rep <- sdreport(obj)
  fxd <- summary(rep, "report")
  fxd_log <- summary(rep, "fixed")
  rdm <- summary(rep, "random")

  lg <- rdm[rownames(rdm) %in% "lg",]
  ll <- rdm[rownames(rdm) %in% "ll",]
  b <- rdm[rownames(rdm) %in% "b", ]
  
  ## build table of ranefs
  nms <- c(gnm, unlist(reTrms$cnms))
  ## get number of random terms
  nrt <- sapply(data.tmb$terms, function(x)
    x$blockSize) %>% sum()
  ret <- matrix(b[, "Estimate"], A, nrt, byrow = TRUE) %>%
    as.data.frame() %>%
    data.frame(unique(reTrms$flist[[1]]), .) %>%
    tbl_df()
  names(ret) <- nms

  ## build table of gamma & lambda estimates
  fitted <- data_frame(
    id = data$id,
    date = data$date,
    gamma = plogis(lg[, 1]),
    gamma.se = lg[, 2],
    lambda = plogis(ll[, 1]),
    lambda.se = ll[, 2]
  )

  if (optim == "nlminb") {
    aic <- 2 * length(opt[["par"]]) + 2 * opt[["objective"]]
    bic <-
      log(nrow(data)) * length(opt[["par"]]) + 2 * opt[["objective"]]
  }
  else {
    aic <- 2 * length(opt[["par"]]) + 2 * opt[["value"]]
    bic <- log(nrow(data)) * length(opt[["par"]]) + 2 * opt[["value"]]
  }

  rownames(fxd)[rownames(fxd) %in% "sigma"] <- c("sigma_long", "sigma_lat", "sigma_depth")
  rownames(fxd)[rownames(fxd) %in% "beta"][1] <- "Intercept"

  ft <- attr(termf, "term.labels")
  rownames(fxd)[rownames(fxd) %in% "beta"] <- ft

  ## FIXME:: need to simplify and organise...
  structure(
    list(
      call = call,
      formula = formula,
      data = data,
      mf = mf,
      fr = fr,
      fitted = fitted,
      par = fxd,
      re = ret,
      tmb = obj,
      opt = opt,
      method = method,
      rep = rep,
      aic = aic,
      bic = bic,
      opt.time = opt.time
    ),
    class = "mpmm"
  )

}

